### PR TITLE
Add remaining link options tests

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
Wrap disposition error in a DetachError when auto-detaching on
disposition error is enabled.

### CODE REVIEW

File: link.go [hunk 1]
```diff
@@ -734,7 +734,7 @@ func (l *link) muxHandleFrame(fr frames.FrameBody) error {
 		//
 		// This isn't ideal, but there isn't a clear better way to handle it.
 		if fr, ok := fr.State.(*encoding.StateRejected); ok && errOnRejectDisposition {
-			return fr.Error
+			return &DetachError{fr.Error}
```
Comment (richardpark-msft) [inline link.go:737]: Why is this a detach error? This is just a failed disposition, right?

Comment (jhendrixMSFT) [reply link.go:737]: We have a setting, [LinkDetachOnDispositionError](https://github.com/Azure/go-amqp/blob/master/link_options.go#L364), that when enabled (true by default), says that disposition errors should detach the link.

Comment (jhendrixMSFT) [reply link.go:737]: In fact, we should also return a `DetachError` from the first `Send` as well when this mode is enabled.

Comment (richardpark-msft) [reply link.go:743]: It's a bit odd, but I think it makes sense.

The actual condition appears to only fire off if we're a _Sender_, which is different than I was thinking (had to actually look at the condition, nothing else appears to indicate that). It's interesting that it's true by default, but I think that's a discussion for another day.

File: link.go [hunk 2]
```diff
@@ -638,11 +638,17 @@ func (l *link) IssueCredit(credit uint32) error {
 	return nil
 }
 
+func (l *link) detachOnRejectDisp() bool {
+	if l.detachOnDispositionError && (l.receiver == nil && (l.ReceiverSettleMode == nil || *l.ReceiverSettleMode == ModeFirst)) {
```
Comment (richardpark-msft) [inline link.go:642]: I know you didn't write this but what is this checking? I guess the receiver settle mode _also_ applies to senders?

Comment (richardpark-msft) [reply link.go:642]: Or is it just trying to ensure that all the 'receiver' fields are set to their default values?

**Review by richardpark-msft — APPROVED** [review]
Looks good. The check for whether we should reject on disposition is odd - you didn't write it, but if you understand it well enough do you mind commenting on the significance of the fields it's checking?

(l.receiver == nil makes total sense).

File: link.go (hunk 2 above)
Comment (jhendrixMSFT) [reply link.go:645]: From the spec.

> When set at the sender this indicates the desired value for the settlement mode at the receiver. When set at the receiver this indicates the actual settlement mode in use. The receiver SHOULD respect the sender's desired settlement mode if the sender initiates the attach exchange and the receiver supports the desired mode.

This check is saying, if the receiver settlement mode (RSM) is unspecified or ModeFirst, we automatically detach when receiving a rejection disposition.  I've added a comment that explains this.

Comment (richardpark-msft) [reply link.go:645]: Ah, that makes a lot of sense. Thanks for that.

**Pull request merged**